From: Daniel Kahn Gillmor <dkg@fifthhorseman.net>
Date: Thu, 30 May 2019 15:16:26 -0400
Subject: protected headers: only put Subject in the legacy-display part

When protecting headers for outbound encrypted e-mail, enigmail
creates a "legacy-display" text/rfc822-headers part.  That part is
intended to be visible for users who read the e-mail in an MUA that is
capable of decryption but not capable of directly handling the
protected headers.

The purpose of the legacy-display part is to make potentially-hidden
aspects of the e-mail visible to these legacy users.

Since the Subject header is protected by stubbing it out externally,
it is appropriate to include it in the legacy-display part.

The other headers, however, are *not* stubbed or stripped externally,
so there is no reason to include them in the legacy-display part.

This change limits the legacy-display part to only the Subject line.
If other headers were stubbed or stripped, then they should be handled
in the same fashion as the Subject header, but at the moment, only the
Subject is protected in this way.

Signed-off-by: Daniel Kahn Gillmor <dkg@fifthhorseman.net>
---
 package/mimeEncrypt.jsm | 14 +-------------
 1 file changed, 1 insertion(+), 13 deletions(-)

diff --git a/package/mimeEncrypt.jsm b/package/mimeEncrypt.jsm
index 7473bab..a04304f 100644
--- a/package/mimeEncrypt.jsm
+++ b/package/mimeEncrypt.jsm
@@ -293,22 +293,10 @@ PgpMimeEncrypt.prototype = {
       }
     };
 
-    // visible headers list
-    let vH = [
-      'from',
-      'to',
-      'subject',
-      'cc'
-    ];
-
     for (let i in h) {
       if (this.msgCompFields[i] && this.msgCompFields[i].length > 0) {
         allHdr += jsmime.headeremitter.emitStructuredHeader(h[i].field, h[i].parser(this.msgCompFields[i]), {});
       }
-
-      if (vH.indexOf(i) >= 0 && this.msgCompFields[i].length > 0) {
-        visibleHdr += jsmime.headeremitter.emitStructuredHeader(h[i].field, h[i].parser(this.msgCompFields[i]), {});
-      }
     }
 
     if (this.enigmailFlags.originalSubject && this.enigmailFlags.originalSubject.length > 0) {
@@ -332,7 +320,7 @@ PgpMimeEncrypt.prototype = {
       allHdr + '\r\n' +
       "--" + this.encHeader + "\r\n";
 
-    if (this.cryptoMode == MIME_ENCRYPTED && this.enigmailFlags.sendFlags & EnigmailConstants.ENCRYPT_HEADERS) {
+    if (this.cryptoMode == MIME_ENCRYPTED && this.enigmailFlags.sendFlags & EnigmailConstants.ENCRYPT_HEADERS && visibleHdr) {
       w += 'Content-Type: text/rfc822-headers; protected-headers="v1"\r\n' +
         'Content-Disposition: inline\r\n\r\n' +
         visibleHdr +
